shell/platform/darwin/macos/framework/Headers/FlutterApplication.h

```diff
@@ -0,0 +1,58 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef FLUTTER_FLUTTERAPPLICATION_H_
+#define FLUTTER_FLUTTERAPPLICATION_H_
+
+#import <Cocoa/Cocoa.h>
+
+/**
+ * A Flutter-specific subclass of NSApplication that overrides |terminate| and
+ * provides an additional |terminateApplication| method so that Flutter can
+ * handle requests for termination in an asynchronous fashion.
+ *
+ * When a call to |terminate| comes in, either from the OS through a Quit menu
+ * item, through the Quit item in the dock context menu, or from the application
+ * itself, a request is sent to the Flutter framework. If that request is
+ * granted, this subclass will (in |terminateApplication|) call
+ * |NSApplication|'s version of |terminate| to proceed with terminating the
+ * application normally by calling |applicationShouldTerminate|, etc.
+ *
+ * If the termination request is denied by the framework, then the application
+ * will continue to execute normally, as if no |terminate| call were made.
+ *
+ * The |FlutterAppDelegate| always returns |NSTerminateNow| from
+ * |applicationShouldTerminate|, since it has already decided by that point that
+ * it should terminate.
+ *
+ * In order for this class to be used in place of |NSApplication|, the
+ * "NSPrincipalClass" entry in the Info.plist for the application must be set to
+ * "FlutterApplication". If it is not, then the application will not be given
+ * the chance to deny a termination request, and calls to requestAppExit on the
+ * engine (from the framework, typically) will simply exit the application
+ * without ceremony.
+ *
+ * If the |NSApp| global isn't of type |FlutterApplication|, a log message will
+ * be printed once in debug mode when the application is first accessed through
+ * the singleton's |sharedApplication|, describing how to fix this.
+ *
+ * Flutter applications are *not* required to inherit from this class.
+ * Developers of custom |NSApplication| subclasses should copy and paste code as
+ * necessary from FlutterApplication.mm.
+ */
+@interface FlutterApplication : NSApplication
+
+/**
+ * FlutterApplication's implementation of |terminate| doesn't terminate the
+ * application: that is left up to the engine, which will call this function if
+ * it decides that termination request is granted, which will start the regular
+ * Cocoa flow for terminating the application, calling
+ * |applicationShouldTermnate|, etc.
+ *
+ * @param(sender) The id of the object requesting the termination, or nil.
+ */
+- (void)terminateApplication:(id)sender;
```

Comment:
I'm fine with the current name, but perhaps this could align to `AppExitType.required`? Perhaps something like `requiredTerminate`?

Reply:
Hmm.  Maybe?  The framework can always send a "required" termination to call it, but Objective-C code might want access.

There is an internal property on the engine that reveals the `terminationHandler`, which could be made public if we want to allow Objective-C code to send a required exit.  It would allow us to do other things when they try and exit if we need to in the future (not sure what those would be, but maybe notifying the framework so it can do work before we call the real terminate? Seems unreliable in the same way that the exiting state was, though.).

Still, add this to a protocol might be a good idea.

Reply:
Ah I think it's sufficient to put it in the header, I just wonder if it's necessary. i.e. can we expose this method just to ourselves and not add any public API for this?

Reply:
Oh, I see, so you're asking if we can make it just an interface declaration with no methods (which we have to have so that they can use it in their Info.plist), and move the method to a private header?

Reply:
Okay, I've done that.